File: src/main/java/com/gitlab/daring/image/command/structure/WatershedMarkerCommand.java
package com.gitlab.daring.image.command.structure;

import com.gitlab.daring.image.command.BaseCommand;
import com.gitlab.daring.image.command.CommandEnv;
import com.gitlab.daring.image.command.parameter.IntParam;
import com.gitlab.daring.image.command.parameter.StringParam;
import org.bytedeco.javacpp.opencv_core.*;

import static com.gitlab.daring.image.util.ImageUtils.smat;
import static org.bytedeco.javacpp.opencv_core.*;
import static org.bytedeco.javacpp.opencv_imgproc.*;

public class WatershedMarkerCommand extends BaseCommand {

    final StringParam markerKey = stringParam("");
    final IntParam segment = intParam(0, "0-10");

    final Mat lm = new Mat();
    final Mat rm = new Mat();

    public WatershedMarkerCommand(String... args) {
        super(args);
    }

    @Override
    public void execute(CommandEnv env) {
        Mat m = env.getMat(markerKey.v);
        threshold(m, m, 128, 255, THRESH_BINARY);
        connectedComponents(m, lm);
        watershed(env.mat, lm);
        max(lm, 0).asMat().convertTo(rm, CV_8U);
        Mat si = smat(segment.v);
        inRange(rm, si, si, rm);
        env.mat = rm;
    }

}

File: src/main/kotlin/com/gitlab/daring/image/command/structure/WatershedMarkerCommand.kt
package com.gitlab.daring.image.command.structure

import com.gitlab.daring.image.command.CommandEnv
import com.gitlab.daring.image.command.KBaseCommand
import com.gitlab.daring.image.util.ImageUtils.smat
import org.bytedeco.javacpp.opencv_core.*
import org.bytedeco.javacpp.opencv_imgproc.*

internal class WatershedMarkerCommand(args: Array<String>) : KBaseCommand(args) {

    val markerKey = stringParam("")
    val segment = intParam(0, "0-10")

    val lm = Mat()
    val rm = Mat()

    override fun execute(env: CommandEnv) {
        val m = env.getMat(markerKey.v)
        threshold(m, m, 128.0, 255.0, THRESH_BINARY)
        connectedComponents(m, lm)
        watershed(env.mat, lm)
        max(lm, 0.0).asMat().convertTo(rm, CV_8U)
        val si = smat(segment.v)
        inRange(rm, si, si, rm)
        env.mat = rm
    }

}
